### ISSUE TITLE
Support standard output formats

### ISSUE THREAD

**Issue by flying-sheep** [python-jsonschema/jsonschema#1008]
The standard draft 2020-12 defines four standard output formats: https://json-schema.org/draft/2020-12/json-schema-core.html#name-output-formatting

- Flag - A boolean which simply indicates the overall validation result with no further details.
- Basic - Provides validation information in a flat list structure.
- Detailed - Provides validation information in a condensed hierarchical structure based on the structure of the schema.
- Verbose - Provides validation information in an uncondensed hierarchical structure that matches the exact structure of the schema.

`jsonschema` should provide ways to emit them and integrate them with its error structure:

- there should be a `Validator` method with a `format` keyword argument that just emits a Python dict conforming to the chosen format
- the ValidationErrors emitted by `iter_errors` should match the errors in the `Basic` format
- the ErrorTree structure should match the `Detailed`/`Verbose` format

**Comment by Julian** [comment]
Hi there! This is certainly welcome/desired, though the implementation is a bit different.

The output formats are actually being restructured heavily still (see https://github.com/json-schema-org/json-schema-spec/pull/1249 and https://github.com/json-schema-org/json-schema-spec/issues/1320) for the next release of the specification.

What this means here is that certainly for now, though implementing them is very much desired and will be done, that the *mechanism* is going to follow more closely the API that exists here and has existed forever -- meaning -- I want this library, in its `ValidationError` [details](https://python-jsonschema.readthedocs.io/en/stable/errors/) to essentially support all of the data needed to *generate* those output formats, or any changes to them.

My quick pass previously was that that's indeed either fully possible or very nearly possible to do already today, simply by taking a bunch of `ValidationError`s and generating the specified JSON for each format. Any gaps here though are 100% stuff that should be added.

Someone (if no one else, eventually me) then should package functions into a package which depends on this one and generates the specific output format (and I'm happy to host+promote that package in this org).

So, a `jsonschema-output2020` package, which, in pseudocode supports:

```python
from jsonschema.validators import Draft2020Validator
import jsonschema_output2020
errors = Draft2020Validator(schema).iter_errors(123)
output = jsonschema_output.detailed(errors=errors)
print(json.dumps(output))
```

is what I believe is the right next step here, along with collecting any missing things from the `ValidationError` API which would support such a thing.

If/when things stabilize upstream, it's possible that the functionality can fully move into this library, though it's something we can independently think about.

> the ValidationErrors emitted by iter_errors should match the errors in the Basic format
> the ErrorTree structure should match the Detailed/Verbose format

These objects have backwards compatibility guarantees, so regardless of anything else, this sort of thing wouldn't change, not without long deprecations.

Hope that clarifies, and very much could use some help if you're interested in giving such a thing a shot.

**Comment by flying-sheep** [comment]
Thank you for the detailed answer. I encountered some problems using `ErrorTree`, so I wasn’t sure how reliable a translation would be (e.g. `ErrorTree` only contains the first error for multiple missing required properties, ignoring all subsequent ones).

Granted, that only applies to `ErrorTree`, not `iter_errors`.

**Comment by Julian** [comment]
Yes `ErrorTree` is completely unrelated (though the issue you mention with it is #442 ).

**Comment by Julian** [comment]
This issue clearly has stagnated anyhow, but given that upstream the specification has essentially put output formats on the backburner and/or relegated them back to being in a sort of pre-alpha status, I'm disinclined to add any support here for them. I still think we definitely *do* want to ensure ValidationErrors contain everything needed to *produce* any output format, and welcome any issues filed about gaps there (e.g. #119) but don't think it's worth investing any effort in supporting generation of "standard" output formats until there's actually a wider used such standard, given how easy it is to produce whatever desired format directly in user code.
